kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/ISO8601/Duration.kt

```diff
@@ -0,0 +1,280 @@
+            command = { mk_Duration(duration.milliseconds % OneSecondDuration.milliseconds) },
+            post = { result -> result < OneSecondDuration }
+        )
+
+        val modMinutes: () -> Duration = function(
+            command = { mk_Duration(duration.milliseconds % OneMinuteDuration.milliseconds) },
+            post = { result -> result < OneMinuteDuration }
+        )
+        val modHours: () -> Duration = function(
+            command = { mk_Duration(duration.milliseconds % OneHourDuration.milliseconds) },
+            post = { result -> result < OneHourDuration }
+        )
+        val modDays: () -> Duration = function(
+            command = { mk_Duration(duration.milliseconds % OneDayDuration.milliseconds) },
+            post = { result -> result < OneDayDuration }
+        )
+
+        private val formatItem: (nat, Char) -> String = function(
+            command = { n, c -> if (n == 0) "" else String.format("%d%s", n, c) }
+        )
+
+        private val formatItemSec: (nat, nat) -> String = function(
+            command = { seconds, milliseconds -> String.format("%d.%03dS", seconds, milliseconds) }
+        )
+
+        val format: () -> String = function<String>(
+            command = {
+                val numDays = duration.functions.toDays().toInt()
+                val timeOfDay = duration.functions.modDays().functions.toTimeAfterFirstTime()
+                val date = formatItem(numDays, 'D')
+                val time = formatItem(timeOfDay.hour, 'H') + formatItem(timeOfDay.minute, 'M') +
+                        if (timeOfDay.millisecond == 0) {
+                            formatItem(timeOfDay.second, 'S')
+                        } else {
+                            formatItemSec(timeOfDay.second, timeOfDay.millisecond)
+                        }
+                if (date == "" && time == "") "PT0S" else "P$date${if (time == "") "" else "T$time"}"
+            }
+        )
+    }
+}
+
+val minDuration: (Set1<Duration>) -> Duration = function(
+    command = { durations -> durations.min() },
+    post = { durations, result -> result in durations && forall(durations) { result <= it } }
+)
+val maxDuration: (Set1<Duration>) -> Duration = function(
+    command = { durations -> durations.max() },
+    post = { durations, result -> result in durations && forall(durations) { result >= it } }
+)
+val sumDuration: (Sequence<Duration>) -> Duration = function(
+    command = { durationSequence -> mk_Duration((seq(durationSequence) { it.milliseconds }).sum()) },
+    post = { durationSequence, result -> forall(durationSequence) { result >= it } }
+)
+
+val durationDiff: (Duration, Duration) -> Duration = function(
+    command = { duration1, duration2 -> mk_Duration(abs(duration1.milliseconds - duration2.milliseconds)) },
+    post = { duration1, duration2, result ->
+        ((duration1 <= duration2) implies (duration1.functions.addDuration(result) == duration2)) &&
```

Comment:
What is going on here? Things like `(duration1.functions.addDuration(result) == duration2)) ` have a natural meaning, it should be clear from reading what that is

Reply:
The function calculates the difference in length between two durations. The post condition checks that the smaller duration plus the difference between the two is the larger duration

Reply:
It's been edited for clarity